be/src/formats/parquet/column_converter.cpp

```diff
@@ -180,63 +180,129 @@ class PrimitiveToDecimalConverter : public ColumnConverter {
     DestPrimitiveType _scale_factor = 1;
 };
 
+// This class is to convert *fixed length* binary to decimal
+// and for fixed length binary in parquet, string data is contiguous,
+// and that's why we can do memcpy 8 bytes without accessing invalid address.
 template <PrimitiveType DestType>
 class BinaryToDecimalConverter : public ColumnConverter {
 public:
     using DecimalType = typename vectorized::RunTimeTypeTraits<DestType>::CppType;
     using ColumnType = typename vectorized::RunTimeTypeTraits<DestType>::ColumnType;
     using DestPrimitiveType = typename vectorized::RunTimeTypeTraits<TYPE_DECIMAL128>::CppType;
-
-    BinaryToDecimalConverter(int32_t src_scale, int32_t dst_scale) {
+    BinaryToDecimalConverter(int32_t src_scale, int32_t dst_scale, int32_t type_length) {
         if (src_scale < dst_scale) {
             _scale_type = DecimalScaleType::kScaleUp;
             _scale_factor = get_scale_factor<DestPrimitiveType>(dst_scale - src_scale);
         } else if (src_scale > dst_scale) {
             _scale_type = DecimalScaleType::kScaleDown;
             _scale_factor = get_scale_factor<DestPrimitiveType>(src_scale - dst_scale);
         }
+        _type_length = type_length;
+    }
+
+    template <int BINSZ, DecimalScaleType scale_type, typename T>
+    void t_convert(size_t size, uint8_t* dst_null_data, uint8_t* src_null_data, DecimalType* dst_data,
+                   const uint8* src_data, bool* has_null) {
+        for (size_t i = 0; i < size; i++) {
+            dst_null_data[i] = src_null_data[i];
+            if (dst_null_data[i]) {
+                *has_null = true;
+                continue;
+            }
+            // When Decimal in parquet is stored in byte arrays, binary and fixed,
+            // the unscaled number must be encoded as two's complement using big-endian byte order.
+
+            DestPrimitiveType unscale = 0;
+            T value = 0;
+            static_assert(BINSZ <= sizeof(value));
+
+            // NOTE(yan): since fixed length binary data is contiguous, with condition check (i+8) < size
+            // we can assure that there is no illegal memory access
+            // UPDATE: bytes are allocated by `RawVectorPad16`, so there will be extra bytes at tail.
+
+            // mempcy 8 bytes to compiler, it's instruction to move 8 bytes from memory to register
+            // and follow actions are all in-register instructions.
+            memcpy(reinterpret_cast<char*>(&value), src_data, sizeof(value));
```

Comment:
```
memcpy(reinterpret_cast<char*>(&value), src_data, sizeof(value));
value = BitUtil::big_endian_to_host(value);
if constexpr(BINSZ < sizeof(value)) {
    value = value >> ((sizeof(value) - BINSZ) * 8);
}
```

Reply:
I don't think we need this condition statement because modern & decent compiler has capbilities of:
- constant propagation & folding
- peephole optimization.

`BINSZ` is constant at compile time, and total expression whill be transformed to 

- `value = value >> ((16 - 16)* 8);` then 
-  `value = value >> 0; ` then
- `value =value` and removed this expression.